### PR TITLE
consider systime's difference between swarm and engine

### LINKED ISSUE #2310: `docker ps` shows incorrect time in STATUS if Swarm and Docker Engine's systime differs
Hi, All,

Today I `docker run` a container against Swarm and immediately run `docker ps`, it shows:
```
~ $  docker run -d ubuntu:14.04 sleep 10000
bfb0964542a425106f8eff04e380887ab99e7bfa4c56ef25daabf78d336ec12c
~ $  docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
bfb0964542a4        ubuntu:14.04        "sleep 10000"       12 hours ago        Up 12 hours                             ubuntu/sharp_mayer
~ $
```

In the field STATUS, it appear `Up 12 hours` which is definitely incorrect. The root cause of this is that the systime of Swarm Host and Docker Host differs. While Swarms use its own time to minus Docker Host time to get the result via https://github.com/docker/swarm/blob/master/cluster/container.go#L46-L73:
```
// FullStateString returns human-readable description of the state
func FullStateString(state *types.ContainerState) string {
	startedAt, _ := time.Parse(time.RFC3339Nano, state.StartedAt)
	finishedAt, _ := time.Parse(time.RFC3339Nano, state.FinishedAt)
	if state.Running {
		if state.Paused {
			return fmt.Sprintf("Up %s (Paused)", units.HumanDuration(time.Now().UTC().Sub(startedAt)))
		}
		if state.Restarting {
			return fmt.Sprintf("Restarting (%d) %s ago", state.ExitCode, units.HumanDuration(time.Now().UTC().Sub(finishedAt)))
		}
		return fmt.Sprintf("Up %s", units.HumanDuration(time.Now().UTC().Sub(startedAt)))
	}

	if state.Dead {
		return "Dead"
	}

	if startedAt.IsZero() {
		return "Created"
	}

	if finishedAt.IsZero() {
		return ""
	}

	return fmt.Sprintf("Exited (%d) %s ago", state.ExitCode, units.HumanDuration(time.Now().UTC().Sub(finishedAt)))
}
```

Swarm uses `time.Now()` of itself, while it does not synchronize with Docker Host.

### PR DESCRIPTION
**Pull request by allencloud**
fixed #2310 

This PR did:
1. add a field **DeltaDuration** in struct Engine;
2. use `swarm systime - docker systime in info api` to fill DeltaDuration
3. when calculate Status and Created time in `docker ps`, use DeltaDuration

Signed-off-by: allencloud <allen.sun@daocloud.io>

### CODE REVIEW

File: cluster/engine.go [hunk 1]
```diff
@@ -491,6 +492,14 @@ func (e *Engine) updateSpecs() error {
 		e.lastError = message
 		return fmt.Errorf(message)
 	}
+
+	if info.SystemTime != "" {
+		engineTime, _ := time.Parse(time.RFC3339Nano, info.SystemTime)
+		e.DeltaDuration = time.Now().UTC().Sub(engineTime)
```
Comment (dongluochen) [inline cluster/engine.go]: This delta is an estimation of time difference between manager and engine with adjustment of delays (Engine response delay + network delay + manager process delay). The adjustment of delays is not a constant. If the servers are sync up on time, this delta might be the source of error. 

Another way to fix is to output time difference as error in log and in `docker ps`, but without adjusting time. 

I think both approaches are ok.

Comment (dongluochen) [inline cluster/engine.go]: I'm thinking if it's better to take a timestamp after line 452, only if time difference is larger than a threshold (1 or 2 seconds), use the DeltaDuration.

File: cluster/container.go [hunk 2]
```diff
@@ -44,17 +44,21 @@ func StateString(state *types.ContainerState) string {
...
 		if state.Paused {
-			return fmt.Sprintf("Up %s (Paused)", units.HumanDuration(time.Now().UTC().Sub(startedAt)))
+			return fmt.Sprintf("Up %s (Paused)", units.HumanDuration(time.Now().UTC().Sub(startedAt.Add(duration))))
 		}
 		if state.Restarting {
-			return fmt.Sprintf("Restarting (%d) %s ago", state.ExitCode, units.HumanDuration(time.Now().UTC().Sub(finishedAt)))
+			return fmt.Sprintf("Restarting (%d) %s ago", state.ExitCode, units.HumanDuration(time.Now().UTC().Sub(finishedAt.Add(duration))))
 		}
-		return fmt.Sprintf("Up %s", units.HumanDuration(time.Now().UTC().Sub(startedAt)))
+		return fmt.Sprintf("Up %s", units.HumanDuration(time.Now().UTC().Sub(startedAt.Add(duration))))
```
Comment (dongluochen) [inline cluster/container.go]: A check is needed here to make sure "time.Now().UTC().Sub(startedAt.Add(duration)))" is not in the future. A container cannot be `up -60 seconds`, i.e., in the future.

Comment (dongluochen) [inline cluster/container.go]: I suspect this wouldn't happen :) But please test.

Comment (allencloud) [inline cluster/container.go]: Yes, it is reasonable. Today I will polish this PR.

File: cluster/engine.go (hunk 1 above)
Comment (allencloud) [inline cluster/engine.go]: I will add the threshold in updateSpecs. So Swarm will not enter an if when receiving a request

**Comment by allencloud** [comment]
This PR needs to change engine_test.go, while I found test in this file is always in a single line and is not so readable. As a result I changed a lot in format of engine_test.go. @nishanttotla 

I added this delta in the raw data Info in updateContainer, so I do not have to change codes in container.go and handler.go. @dongluochen

**Comment by allencloud** [comment]
ping @dongluochen @nishanttotla @vieux @jimmyxian

**Comment by dongluochen** [comment]
@allencloud Please rebase.

**Comment by allencloud** [comment]
@dongluochen Rebased. PTAL

**Comment by dongluochen** [comment]
LGTM

**Comment by abronan** [comment]
LGTM

**Pull request merged**